### PR TITLE
Rebase 20160411

### PR DESCRIPTION
**Pull request by deads2k**
Before rebasing on master, all code compiled, and all tests passed.

 - [ ] enable third party resources.  @liggitt wants API changes first and we probably need a full cached discovery client to avoid tons of round trips.
 - [ ] migrate to the new deep copy generator.  The old one is still work for us, but it isn't being improved.  The new deep copy generator had different assumptions about packages.  It wasn't a drop-in.
 - [ ] enable recursive directory options in our cli where appropriate.
 - [ ] @liggitt wants to have a single authorization attributes interface for both kube and origin.  That means changing the origin sig in a non-trivial way.

@pweil- Ready for review for whichever victim you choose.

### CODE REVIEW

**Comment by deads2k** [comment]
[test]

**Comment by deads2k** [comment]
@liggitt @ncdc @dgoodwin @sosiouxme @bparees 

Each of you has a commit with your name/ID in all caps.  Please look at the those commits and review or provide patches as appropriate.

**Comment by deads2k** [comment]
@openshift/networking This rebase brought in a new method required for networking plugins.  To make progress, I simply stubbed this in under the `UPSTREAM: openshift/openshift-sdn: <drop>: BREAK OVS PLUGINgit status!` commit.  I don't know the proper fix, can you please provide a patch?

File: pkg/build/registry/buildconfiginstantiate/rest.go [hunk 1]
```diff
@@ -209,7 +210,7 @@ func (h *binaryInstantiateHandler) handle(r io.Reader) (runtime.Object, error) {
 	if err != nil {
 		return nil, errors.NewInternalError(fmt.Errorf("unable to connect to server: %v", err))
 	}
-	if err := exec.Stream(r, nil, nil, false); err != nil {
+	if err := exec.Stream([]string{kubeletremotecommand.StreamProtocolV2Name, kubeletremotecommand.StreamProtocolV1Name}, r, nil, nil, false); err != nil {
```
Comment (ncdc) [inline pkg/build/registry/buildconfiginstantiate/rest.go]: You can use `kubeletremotecommand.SupportedStreamingProtocols`

Comment (deads2k) [inline pkg/build/registry/buildconfiginstantiate/rest.go]: > You can use kubeletremotecommand.SupportedStreamingProtocols

Done.

**Comment by pweil-** [comment]
@miminar - you're up to bat on this one since @kargakis and @soltysh are out most of this week.  I'll take a look too.

@legionus @sgallagher - FYI - a rebase in all its glory.  One day the rebase card will have your name on it :smiling_imp:

**Comment by eparis** [comment]
@deads2k when is this hoping to land?

**Comment by sosiouxme** [comment]
looking at it

**Comment by danwinship** [comment]
@deads2k it looks like the ovs plugin Capabilities() method should be:

    return utilsets.NewInt(knetwork.NET_PLUGIN_CAPABILITY_SHAPING)

**Comment by deads2k** [comment]
@eparis immediately after 1.2 is tagged before the queue reopens.

**Comment by dgoodwin** [comment]
Regarding 9a6a6ed0781e40132deb594bfb2d3481138cb9c4 the upstream commit which changed the interface was https://github.com/kubernetes/kubernetes/commit/79012f6d and it does look like a pretty straightforward rename. 

In a perfect world we should probably remove the "builder" from all the variable names/structs/comments, so I will submit a PR to your branch momentarily @deads2k.

**Comment by deads2k** [comment]
> @deads2k it looks like the ovs plugin Capabilities() method should be:
> 
> return utilsets.NewInt(knetwork.NET_PLUGIN_CAPABILITY_SHAPING)

done

**Comment by deads2k** [comment]
@jwforres known test-assets problem or have I borked something: https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin_integration/56/consoleText ?

File: hack/test-go.sh [hunk 2]
```diff
@@ -186,10 +186,11 @@ else
             -o -path "${kubernetes_path}/pkg/api/v1"                                                  \
             -o -path "${kubernetes_path}/test/e2e"                                                    \
             -o -path "${kubernetes_path}/cmd/libs/go2idl/client-gen/testoutput/testgroup/unversioned" \
+            -o -path "${kubernetes_path}/pkg/storage/etcd3" \
           \) -prune                                                                                   \
         \) -name '*_test.go' | xargs -n1 dirname | sort -u | xargs -n1 printf "${OS_GO_PACKAGE}/%s\n")"
 
-        test_packages="${test_packages} ${optional_kubernetes_packages}"
+        test_packages="${optional_kubernetes_packages}"
```
Comment (stevekuznetsov) [inline hack/test-go.sh]: this disables Kube API tests as well

File: pkg/client/scale.go [hunk 3]
```diff
@@ -35,6 +36,26 @@ func NewDelegatingScaleNamespacer(dcNamespacer DeploymentConfigsNamespacer, sNam
 	}
 }
 
+type delegatingScaleNamespacer2 struct {
```
Comment (miminar) [inline pkg/client/scale.go]: The higher the number, the bettter the namespacer? Do we really need two?

File: pkg/cmd/server/kubernetes/master_config.go [hunk 4]
```diff
@@ -338,20 +338,24 @@ func BuildKubernetesMasterConfig(options configapi.MasterConfig, requestContextM
 }
 
 // getAPIGroupVersionOverrides builds the overrides in the format expected by master.Config.APIGroupVersionOverrides
```
Comment (miminar) [inline pkg/cmd/server/kubernetes/master_config.go]: `s/Config.APIGroupVersionOverrides/Config.APIResourceConfigSource/`

File: pkg/cmd/server/kubernetes/master_config.go [hunk 5]
```diff
@@ -338,20 +338,24 @@ func BuildKubernetesMasterConfig(options configapi.MasterConfig, requestContextM
...
-func getAPIGroupVersionOverrides(options configapi.MasterConfig) map[string]genericapiserver.APIGroupVersionOverride {
-	apiGroupVersionOverrides := map[string]genericapiserver.APIGroupVersionOverride{}
+func getAPIGroupVersionOverrides(options configapi.MasterConfig) genericapiserver.APIResourceConfigSource {
```
Comment (miminar) [inline pkg/cmd/server/kubernetes/master_config.go]: Shouldn't this function be renamed as well?

File: pkg/authorization/authorizer/adapter/attributes.go [hunk 6]
```diff
@@ -70,6 +77,26 @@ func (a AdapterAttributes) GetNamespace() string {
 	return a.namespace
 }
 
+func (a AdapterAttributes) GetName() string {
+	return a.AuthorizationAttributes.GetResourceName()
```
Comment (miminar) [inline pkg/authorization/authorizer/adapter/attributes.go]: Why [adding `name`](https://github.com/openshift/origin/pull/8484/commits/9cc4e772219d781eba1f6f682a963046a5b0b3b6#diff-d32078125722ccc2c1dddd974cae6216R20) attribute then?

File: pkg/authorization/authorizer/adapter/attributes.go [hunk 7]
```diff
@@ -70,6 +77,26 @@ func (a AdapterAttributes) GetNamespace() string {
...
+
+func (a AdapterAttributes) GetSubresource() string {
+	tokens := strings.SplitN(a.AuthorizationAttributes.GetResource(), "/", 2)
+	if len(tokens) != 2 {
+		return ""
+	}
+	return tokens[1]
+}
+
+func (a AdapterAttributes) GetResource() string {
```
Comment (miminar) [inline pkg/authorization/authorizer/adapter/attributes.go]: Godocs.

File: hack/test-go.sh (hunk 2 above)
Comment (deads2k) [inline hack/test-go.sh]: > this disables Kube API tests as well

accident, reverting

Comment (deads2k) [inline hack/test-go.sh]: > accident, reverting

Also, how about a "test kube only" mode?

File: pkg/client/scale.go (hunk 3 above)
Comment (deads2k) [inline pkg/client/scale.go]: > The higher the number, the bettter the namespacer? Do we really need two?

Casting problems.  We might.  I'll see if I can tweak it.

File: pkg/authorization/authorizer/adapter/attributes.go (hunk 6 above)
Comment (deads2k) [inline pkg/authorization/authorizer/adapter/attributes.go]: > Why adding name attribute then?

removed

File: pkg/authorization/authorizer/adapter/attributes.go (hunk 7 above)
Comment (deads2k) [inline pkg/authorization/authorizer/adapter/attributes.go]: > Godocs.

matches interface, the interface is doc'ed

**Comment by deads2k** [comment]
comments addressed and rebased

**Comment by stevekuznetsov** [comment]
@deads2k "test only Kube" mode seems useful for you in the rebase but nobody else using `hack/test-go.sh` so I suggest using a `DRY_RUN=true TEST_KUBE=true hack/test-go.sh | sed 'grab the lines you need' | hack/test-go.sh`

**Comment by deads2k** [comment]
> @deads2k "test only Kube" mode seems useful for you in the rebase but nobody else using hack/test-go.sh so I suggest using a DRY_RUN=true TEST_KUBE=true hack/test-go.sh | sed 'grab the lines you need' | hack/test-go.sh

You probably meant to say, "useful to anyone doing an upstream pick who ran `test-go.sh` and then realized that he forgot TEST_KUBE=1"....

**Comment by pweil-** [comment]
if only we had a makefile that would set this so we didn't have to remember :)

**Comment by stevekuznetsov** [comment]
> and then realized that he forgot `TEST_KUBE=1`

I'm not sure *another* mode of operation would help a user if they aren't solid on the current API :)

**Comment by deads2k** [comment]
Ha, green!  merge it quick! :)

**Comment by miminar** [comment]
LGTM after a very mild and amateurish analysis

File: pkg/cmd/server/etcd/etcdserver/server.go [hunk 8]
```diff
@@ -55,11 +55,6 @@ func startEtcd(cfg *config) (<-chan struct{}, error) {
 		return nil, fmt.Errorf("error setting up initial cluster: %v", err)
 	}
```
Comment (liggitt) [inline pkg/cmd/server/etcd/etcdserver/server.go]: @smarterclayton can you review this part?

File: pkg/authorization/authorizer/adapter/attributes.go [hunk 9]
```diff
@@ -70,6 +76,26 @@ func (a AdapterAttributes) GetNamespace() string {
...
+	tokens := strings.SplitN(a.AuthorizationAttributes.GetResource(), "/", 2)
```
Comment (liggitt) [inline pkg/authorization/authorizer/adapter/attributes.go]: this seems like it propagates the bug... `AdapterAttributes` satisfying `oauthorizer.AuthorizationAttributes` with a `GetResource` impl that does not include subresource is scary

we should update `oauthorizer.AuthorizationAttributes` to also have a GetSubresource(), and update `node_auth.go` to set resource and subresource correctly

Comment (deads2k) [inline pkg/authorization/authorizer/adapter/attributes.go]: Not now.  Eliminated it satisfying both interfaces.  Perfect round tripping
is still supported on the exact type.  It should never have been anonymous
and it should probably have never tried to satisfy both interfaces.

**Comment by deads2k** [comment]
@liggitt @pweil- I'd liked this merged after we tag and **before** master re-opens.  Let's prioritize reviews if 1.2 is close.

**Comment by soltysh** [comment]
I'll reserve tomorrow for a rewiew in that case.

**Comment by kargakis** [comment]
> I'll reserve tomorrow for a review in that case.

me too

**Comment by smarterclayton** [comment]
We're probably going to strongly control the merge order post fork to
prioritize the key things.

On Wed, Apr 20, 2016 at 11:44 AM, Michail Kargakis <notifications@github.com

File: pkg/cmd/server/etcd/etcdserver/server.go (hunk 8 above)
Comment (smarterclayton) [inline pkg/cmd/server/etcd/etcdserver/server.go]: Why was this removed?

Comment (smarterclayton) [inline pkg/cmd/server/etcd/etcdserver/server.go]: Not a field anymore?

File: pkg/build/registry/buildconfiginstantiate/rest.go [hunk 10]
```diff
@@ -209,7 +210,7 @@ func (h *binaryInstantiateHandler) handle(r io.Reader) (runtime.Object, error) {
 	if err != nil {
 		return nil, errors.NewInternalError(fmt.Errorf("unable to connect to server: %v", err))
 	}
-	if err := exec.Stream(r, nil, nil, false); err != nil {
+	if err := exec.Stream(kubeletremotecommand.SupportedStreamingProtocols, r, nil, nil, false); err != nil {
```
Comment (smarterclayton) [inline pkg/build/registry/buildconfiginstantiate/rest.go]: @ncdc is there a backcompat implication?

File: pkg/volume/emptydir/empty_dir_quota.go [hunk 11]
```diff
@@ -8,25 +8,25 @@ import (
 )
 
 var _ volume.VolumePlugin = &EmptyDirQuotaPlugin{}
-var _ volume.Builder = &emptyDirQuotaBuilder{}
+var _ volume.Mounter = &emptyDirQuotaMounter{}
```
Comment (smarterclayton) [inline pkg/volume/emptydir/empty_dir_quota.go]: @dgoodwin please review this change

File: pkg/cmd/util/clientcmd/factory.go [hunk 12]
```diff
@@ -170,7 +170,7 @@ func NewFactory(clientConfig kclientcmd.ClientConfig) *Factory {
 		clients:               clients,
 	}
 
-	w.Object = func() (meta.RESTMapper, runtime.ObjectTyper) {
+	w.Object = func(bool) (meta.RESTMapper, runtime.ObjectTyper) {
```
Comment (smarterclayton) [inline pkg/cmd/util/clientcmd/factory.go]: We're going to have to start enabling 3rd party resources soon.

File: Godeps/_workspace/src/k8s.io/kubernetes/pkg/runtime/deep_copy_generator.go [hunk 13]
```diff
@@ -0,0 +1,609 @@
+/*
```
Comment (smarterclayton) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/runtime/deep_copy_generator.go]: Why?

File: pkg/build/registry/buildconfiginstantiate/rest.go (hunk 10 above)
Comment (ncdc) [inline pkg/build/registry/buildconfiginstantiate/rest.go]: It's backwards compatible

**Comment by deads2k** [comment]
Image import.  re[test]

**Comment by deads2k** [comment]
> We're probably going to strongly control the merge order post fork to
prioritize the key things.

Unless we have concurrent hot fixes, I'd like this next.

File: pkg/volume/emptydir/empty_dir_quota.go (hunk 11 above)
Comment (dgoodwin) [inline pkg/volume/emptydir/empty_dir_quota.go]: Already did above when deads looped me in. Checked the upstream change, just a simple rename so looks good.

**Comment by smarterclayton** [comment]
Once this is ready create a PR to smarterclayton:post_12

File: pkg/cmd/util/clientcmd/factory.go (hunk 12 above)
Comment (deads2k) [inline pkg/cmd/util/clientcmd/factory.go]: > We're going to have to start enabling 3rd party resources soon.

Not during the rebase and preferably not like they've done.  We should make a complete caching discovery client upstream, the right code to use a cache unless it fails on an unknown type.  A discovery `RESTMapper` ought not be hard to build either.

File: Godeps/_workspace/src/k8s.io/kubernetes/pkg/runtime/deep_copy_generator.go (hunk 13 above)
Comment (deads2k) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/runtime/deep_copy_generator.go]: > Why?

At the time?  Time.  The new deep copy generator had multiple bad assumptions about package availability, the old deep copy generator worked, and 1.2 was going to be done two days after I finished this so I wanted reviews, merge, then fix-up.

Comment (smarterclayton) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/runtime/deep_copy_generator.go]: Is the follow up issue list linked?

File: pkg/cmd/cli/cmd/debug.go [hunk 14]
```diff
@@ -199,7 +199,7 @@ func (o *DebugOptions) Complete(cmd *cobra.Command, f *clientcmd.Factory, args [
 		ResourceTypeOrNameArgs(true, resources...).
 		Flatten()
 	if len(o.Filename) > 0 {
-		b.FilenameParam(explicit, o.Filename)
+		b.FilenameParam(explicit, false, o.Filename)
```
Comment (kargakis) [inline pkg/cmd/cli/cmd/debug.go]: Is there an issue to enable these?

File: pkg/client/scale.go (hunk 3 above)
Comment (soltysh) [inline pkg/client/scale.go]: I'll re-rise the issue, I don't see the other being used anywhere else.

File: hack/test-go.sh [hunk 15]
```diff
@@ -186,6 +186,7 @@ else
             -o -path "${kubernetes_path}/pkg/api/v1"                                                  \
             -o -path "${kubernetes_path}/test/e2e"                                                    \
             -o -path "${kubernetes_path}/cmd/libs/go2idl/client-gen/testoutput/testgroup/unversioned" \
+            -o -path "${kubernetes_path}/pkg/storage/etcd3"                                           \
```
Comment (soltysh) [inline hack/test-go.sh]: We don't need it right now, since it's not finished, from what I see. But I'd suggest marking this commit as `<drop>` for next rebase.

Comment (kargakis) [inline hack/test-go.sh]: Mind commenting on why are these disabled? Actually, each suite should be followed by such a comment.

File: pkg/cmd/util/clientcmd/factory.go (hunk 12 above)
Comment (kargakis) [inline pkg/cmd/util/clientcmd/factory.go]: follow-up issue?

File: pkg/cmd/cli/cmd/debug.go (hunk 14 above)
Comment (soltysh) [inline pkg/cmd/cli/cmd/debug.go]: Looking at the affected files, I doubt we need recursive file visitor for these commands, at least for now.

**Comment by soltysh** [comment]
I left a few comments, otherwise I'm giving LGTM.

File: hack/test-go.sh (hunk 15 above)
Comment (deads2k) [inline hack/test-go.sh]: > Mind commenting on why are these disabled? Actually, each suite should be followed by such a comment.

Flaky upstream and we don't support etcd3.  Comment added and commit is marked drop.

File: pkg/cmd/server/etcd/etcdserver/server.go (hunk 8 above)
Comment (deads2k) [inline pkg/cmd/server/etcd/etcdserver/server.go]: > Not a field anymore?

Yeah, you can no longer pass a transport.

File: pkg/client/scale.go (hunk 3 above)
Comment (deads2k) [inline pkg/client/scale.go]: > I'll re-rise the issue, I don't see the other being used anywhere else.

done

**Comment by deads2k** [comment]
Updated the checklist at the top, fixed up, and responded to comments.

@smarterclayton I'd like the two LIGGITT reviews signed off on before squashing down and opening against your branch.

**Comment by deads2k** [comment]
well, not to worry.  This has the same image import flake we have in master.

**Comment by smarterclayton** [comment]
SDN had to merge for a p0

On Apr 22, 2016, at 7:10 PM, OpenShift Bot <notifications@github.com> wrote:

Origin Action Required: Pull request cannot be automatically merged, please
rebase your branch from latest HEAD and push again

—
You are receiving this because you were mentioned.
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/8484#issuecomment-213620792>

**Comment by deads2k** [comment]
rebased and squashed down

**Comment by pweil-** [comment]
LGTM

**Comment by smarterclayton** [comment]
After I cut 1.2.0-rc2 this is first in my "i will merge these in order".

**Comment by smarterclayton** [comment]
@soltysh pull through is still failing, can you help me track down where
the retry might be needed?  Possibly in the write path of the ServeBlob
code (it might be succeeding once and then failing later)?  Or we may be
exceeding our retries?
https://ci.openshift.redhat.com/jenkins/job/test_pr_origin/3221/

**Comment by deads2k** [comment]
re[test]

**Comment by deads2k** [comment]
re[test]

**Comment by smarterclayton** [comment]
[merge] for 1.3.0 and NO, the queue is not open yet for other merges
